shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java

```diff
@@ -64,9 +66,7 @@ public static DatabaseType getProtocolType(final Map<String, ? extends DatabaseC
         if (configuredDatabaseType.isPresent()) {
             return configuredDatabaseType.get();
         }
-        Collection<DataSource> dataSources = databaseConfigs.values().stream()
-                .filter(each -> !each.getDataSources().isEmpty()).findFirst().map(optional -> optional.getDataSources().values()).orElseGet(Collections::emptyList);
-        return getDatabaseType(dataSources);
+        return getDatabaseType(DataSourceStateManager.getInstance().getEnabledDataSources(databaseConfigs));
```

Comment:
Why is the databaseName parameter not needed here?

Reply:
You are right.I shouldn't manage logic other than data source state.I will change the api.